consensus/state.go

```diff
@@ -1642,21 +1642,17 @@ func (cs *ConsensusState) addVote(vote *types.Vote, peerID p2p.ID) (added bool,
 		precommits := cs.Votes.Precommits(vote.Round)
 		cs.Logger.Info("Added to precommit", "vote", vote, "precommits", precommits.StringShort())
 		blockID, ok := precommits.TwoThirdsMajority()
-		if ok {
-			if len(blockID.Hash) == 0 {
-				cs.enterNewRound(height, vote.Round+1)
-			} else {
-				cs.enterNewRound(height, vote.Round)
-				cs.enterPrecommit(height, vote.Round)
-				cs.enterCommit(height, vote.Round)
-
-				if cs.config.SkipTimeoutCommit && precommits.HasAll() {
-					// if we have all the votes now,
-					// go straight to new round (skip timeout commit)
-					// cs.scheduleTimeout(time.Duration(0), cs.Height, 0, cstypes.RoundStepNewHeight)
-					cs.enterNewRound(cs.Height, 0)
-				}
+		if ok && len(blockID.Hash) != 0 {
```

Comment:
If `len(blockID.Hash) == 0`, don't we need to explicitly `enterPrecommitWait`?

Reply:
We should always enterPrecommitWait upon receiving 2f+1 any precommits. We don't need to treat 2f+1 Precommit nil as a special case. We do it here: https://github.com/tendermint/tendermint/blob/bb8411b600eb71489baf008a646f31db30fcd642/consensus/state.go#L1657-L1660. Only in case we have 2f+1 precommits for some block we go straight to commit and we don't need to trigger timeout.

Reply:
Ok right, it's taken care of by the `else if`. The first condition is only for `+2/3 majority for a block hash`. If there's a +2/3 majority for nil, it will be treated by the second condition. Thanks.